cmd/api/handlers-queries.go

```diff
@@ -0,0 +1,174 @@
+package main
+
+import (
+	"encoding/json"
+	"log"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/jmpsec/osctrl/pkg/queries"
+	"github.com/jmpsec/osctrl/pkg/settings"
+	"github.com/jmpsec/osctrl/pkg/utils"
+)
+
+// GET Handler to return a single query in JSON
+func apiQueryShowHandler(w http.ResponseWriter, r *http.Request) {
+	incMetric(metricAPIReq)
+	utils.DebugHTTPDump(r, settingsmgr.DebugHTTP(settings.ServiceAPI), false)
+	vars := mux.Vars(r)
+	// Extract name
+	name, ok := vars["name"]
+	if !ok {
+		incMetric(metricAPIErr)
+		apiErrorResponse(w, "error getting name", nil)
+		return
+	}
+	// Get query by name
+	query, err := queriesmgr.Get(name)
+	if err != nil {
+		incMetric(metricAPIErr)
+		if err.Error() == "record not found" {
+			log.Printf("query not found: %s", name)
+			apiHTTPResponse(w, JSONApplicationUTF8, http.StatusNotFound, ApiErrorResponse{Error: "query not found"})
+		} else {
+			apiErrorResponse(w, "error getting query", err)
+		}
+		return
+	}
+	// Serialize and serve JSON
+	apiHTTPResponse(w, JSONApplicationUTF8, http.StatusOK, query)
+	incMetric(metricAPIOK)
+}
+
+// POST Handler to run a query
+func apiQueriesRunHandler(w http.ResponseWriter, r *http.Request) {
+	incMetric(metricAPIReq)
+	utils.DebugHTTPDump(r, settingsmgr.DebugHTTP(settings.ServiceAPI), false)
+	var q DistributedQueryRequest
+	// Parse request JSON body
+	if err := json.NewDecoder(r.Body).Decode(&q); err != nil {
+		incMetric(metricAPIErr)
+		apiErrorResponse(w, "error parsing POST body", err)
+		return
+	}
+	// FIXME check validity of query
+	// Query can not be empty
+	if q.Query == "" {
+		apiErrorResponse(w, "query can not be empty", nil)
+		return
+	}
+	// Prepare and create new query
+	queryName := "query_" + generateQueryName()
+	newQuery := queries.DistributedQuery{
+		Query:      q.Query,
+		Name:       queryName,
+		Creator:    "API",
+		Expected:   0,
+		Executions: 0,
+		Active:     true,
+		Completed:  false,
+		Deleted:    false,
+		Repeat:     0,
```

Comment:
I'm not sure how this value should work. In my eyes a distributed query should run only once. If it runs anymore times than that, then when you fetch something by `queryName` what does that actually mean? Is it original data? The most recent data? Both?

Reply:
This is some legacy stuff, that I thought about having distributed queries in a way that can be repeated, but then again, those queries should be scheduled. I can probably clean this up since it is dead code...